File: services/agent-registry/src/store.ts
import type { Pool } from 'pg';
import type { AgentConfigRecord } from './types';

/** sys_agent_config 数据访问抽象（生产=Postgres，测试=内存）。 */
export interface AgentConfigStore {
  list(): Promise<AgentConfigRecord[]>;
  findByAgentId(agentId: string): Promise<AgentConfigRecord | undefined>;
  create(record: Omit<AgentConfigRecord, 'id' | 'createdAt' | 'updatedAt'>): Promise<AgentConfigRecord>;
  update(agentId: string, patch: Partial<AgentConfigRecord>): Promise<AgentConfigRecord | undefined>;
  delete(agentId: string): Promise<boolean>;
}

interface AgentConfigRow {
  id: string;
  agent_id: string;
  name: string;
  label: string;
  description: string | null;
  system_prompt: string | null;
  mdl: string;
  db_connection_enc: string;
  status: string;
  last_error: string | null;
  created_at: string;
  updated_at: string;
}

function rowToRecord(row: AgentConfigRow): AgentConfigRecord {
  return {
    id: row.id,
    agentId: row.agent_id,
    name: row.name,
    label: row.label,
    description: row.description,
    systemPrompt: row.system_prompt,
    mdl: row.mdl,
    dbConnectionEnc: row.db_connection_enc,
    status: row.status,
    lastError: row.last_error,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
}

const SELECT_COLUMNS = `
  id, agent_id, name, label, description, system_prompt, mdl,
  db_connection_enc, status, last_error, created_at, updated_at
`;

/** PostgreSQL 实现：读写 sys_agent_config 表。 */
export class PostgresAgentConfigStore implements AgentConfigStore {
  constructor(private readonly pool: Pool) {}

  async list(): Promise<AgentConfigRecord[]> {
    const result = await this.pool.query<AgentConfigRow>(
      `SELECT ${SELECT_COLUMNS} FROM sys_agent_config ORDER BY created_at ASC`,
    );
    return result.rows.map(rowToRecord);
  }

  async findByAgentId(agentId: string): Promise<AgentConfigRecord | undefined> {
    const result = await this.pool.query<AgentConfigRow>(
      `SELECT ${SELECT_COLUMNS} FROM sys_agent_config WHERE agent_id = $1`,
      [agentId],
    );
    return result.rows[0] ? rowToRecord(result.rows[0]) : undefined;
  }

  async create(record: Omit<AgentConfigRecord, 'id' | 'createdAt' | 'updatedAt'>): Promise<AgentConfigRecord> {
    const result = await this.pool.query<AgentConfigRow>(
      `INSERT INTO sys_agent_config
        (agent_id, name, label, description, system_prompt, mdl, db_connection_enc, status, last_error)
       VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)
       RETURNING ${SELECT_COLUMNS}`,
      [
        record.agentId,
        record.name,
        record.label,
        record.description,
        record.systemPrompt,
        record.mdl,
        record.dbConnectionEnc,
        record.status,
        record.lastError,
      ],
    );
    return rowToRecord(result.rows[0]!);
  }

  async update(agentId: string, patch: Partial<AgentConfigRecord>): Promise<AgentConfigRecord | undefined> {
    const sets: string[] = [];
    const values: unknown[] = [];
    const fieldMap: Record<string, string> = {
      name: 'name',
      label: 'label',
      description: 'description',
      systemPrompt: 'system_prompt',
      mdl: 'mdl',
      dbConnectionEnc: 'db_connection_enc',
      status: 'status',
      lastError: 'last_error',
    };
    for (const [key, column] of Object.entries(fieldMap)) {
      if (patch[key as keyof AgentConfigRecord] !== undefined) {
        values.push(patch[key as keyof AgentConfigRecord]);
        sets.push(`${column} = $${values.length}`);
      }
    }
    if (sets.length === 0) {
      return this.findByAgentId(agentId);
    }
    values.push(agentId);
    const result = await this.pool.query<AgentConfigRow>(
      `UPDATE sys_agent_config SET ${sets.join(', ')}, updated_at = now()
       WHERE agent_id = $${values.length}
       RETURNING ${SELECT_COLUMNS}`,
      values,
    );
    return result.rows[0] ? rowToRecord(result.rows[0]) : undefined;
  }

  async delete(agentId: string): Promise<boolean> {
    const result = await this.pool.query(
      'DELETE FROM sys_agent_config WHERE agent_id = $1',
      [agentId],
    );
    return (result.rowCount ?? 0) > 0;
  }
}

/** 内存实现（测试/演示用）。 */
export class InMemoryAgentConfigStore implements AgentConfigStore {
  private readonly records = new Map<string, AgentConfigRecord>();
  private seq = 0;

  async list(): Promise<AgentConfigRecord[]> {
    return [...this.records.values()];
  }

  async findByAgentId(agentId: string): Promise<AgentConfigRecord | undefined> {
    return this.records.get(agentId);
  }

  async create(record: Omit<AgentConfigRecord, 'id' | 'createdAt' | 'updatedAt'>): Promise<AgentConfigRecord> {
    const now = new Date().toISOString();
    const full: AgentConfigRecord = {
      ...record,
      id: `mem-${++this.seq}`,
      createdAt: now,
      updatedAt: now,
    };
    this.records.set(record.agentId, full);
    return full;
  }

  async update(agentId: string, patch: Partial<AgentConfigRecord>): Promise<AgentConfigRecord | undefined> {
    const current = this.records.get(agentId);
    if (!current) return undefined;
    const updated: AgentConfigRecord = { ...current, ...patch, agentId, updatedAt: new Date().toISOString() };
    this.records.set(agentId, updated);
    return updated;
  }

  async delete(agentId: string): Promise<boolean> {
    return this.records.delete(agentId);
  }
}

File: backend/app/registry/store.py
"""sys_agent_config 数据访问（对应 TS services/agent-registry/src/store.ts）。

project_json 列存 WrenAI 工程序列化 JSON（用户注册时提交）。
"""
from __future__ import annotations

from typing import Any

import asyncpg


class AgentConfigStore:
    def __init__(self, pool: asyncpg.Pool):
        self._pool = pool

    async def list(self) -> list[dict[str, Any]]:
        async with self._pool.acquire() as conn:
            rows = await conn.fetch(
                "SELECT id, agent_id, name, label, description, system_prompt, "
                "project_json, db_connection_enc, status, last_error, owner_id, "
                "created_at, updated_at FROM sys_agent_config ORDER BY created_at ASC"
            )
        return [self._row_to_record(r) for r in rows]

    async def find_by_agent_id(self, agent_id: str) -> dict[str, Any] | None:
        async with self._pool.acquire() as conn:
            row = await conn.fetchrow(
                "SELECT id, agent_id, name, label, description, system_prompt, "
                "project_json, db_connection_enc, status, last_error, owner_id, "
                "created_at, updated_at FROM sys_agent_config WHERE agent_id = $1",
                agent_id,
            )
        return self._row_to_record(row) if row else None

    async def create(self, record: dict[str, Any]) -> dict[str, Any]:
        async with self._pool.acquire() as conn:
            row = await conn.fetchrow(
                "INSERT INTO sys_agent_config "
                "(agent_id, name, label, description, system_prompt, project_json, "
                "db_connection_enc, status, last_error, owner_id) "
                "VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10) "
                "RETURNING id, agent_id, name, label, description, system_prompt, "
                "project_json, db_connection_enc, status, last_error, owner_id, "
                "created_at, updated_at",
                record["agentId"], record["name"], record["label"],
                record.get("description"), record.get("systemPrompt"),
                record["projectJson"], record["dbConnectionEnc"],
                record.get("status", "enabled"), record.get("lastError"),
                record.get("ownerId"),
            )
        return self._row_to_record(row)

    async def update(self, agent_id: str, patch: dict[str, Any]) -> dict[str, Any] | None:
        sets: list[str] = []
        values: list[Any] = []
        field_map = {
            "name": "name", "label": "label", "description": "description",
            "systemPrompt": "system_prompt", "projectJson": "project_json",
            "dbConnectionEnc": "db_connection_enc", "status": "status",
            "lastError": "last_error", "ownerId": "owner_id",
        }
        for key, column in field_map.items():
            if key in patch:
                values.append(patch[key])
                sets.append(f"{column} = ${len(values)}")
        if not sets:
            return await self.find_by_agent_id(agent_id)
        values.append(agent_id)
        async with self._pool.acquire() as conn:
            row = await conn.fetchrow(
                f"UPDATE sys_agent_config SET {', '.join(sets)}, updated_at = now() "
                f"WHERE agent_id = ${len(values)} RETURNING id, agent_id, name, label, "
                "description, system_prompt, project_json, db_connection_enc, status, "
                "last_error, owner_id, created_at, updated_at",
                *values,
            )
        return self._row_to_record(row) if row else None

    async def delete(self, agent_id: str) -> bool:
        async with self._pool.acquire() as conn:
            result = await conn.execute("DELETE FROM sys_agent_config WHERE agent_id = $1", agent_id)
        return "DELETE 1" in result

    @staticmethod
    def _row_to_record(row: asyncpg.Record) -> dict[str, Any]:
        return {
            "id": str(row["id"]),
            "agentId": row["agent_id"],
            "name": row["name"],
            "label": row["label"],
            "description": row["description"],
            "systemPrompt": row["system_prompt"],
            "projectJson": row["project_json"],
            "dbConnectionEnc": row["db_connection_enc"],
            "status": row["status"],
            "lastError": row["last_error"],
            "ownerId": row["owner_id"],
            "createdAt": row["created_at"].isoformat() if row["created_at"] else None,
            "updatedAt": row["updated_at"].isoformat() if row["updated_at"] else None,
        }
